File: src/lib/api/quicksdk.ts
/**
 * QuickSDK 流水库 API
 */

import { ApiError, apiGet } from '@/lib/api/client.ts'

const PATH = '/api/quicksdk'

export type QuickSdkRdLineSuggestion = {
  game_name: string
  settlement_month: string | null
  row_count: number
  channel_count: number
  source_game_count: number
  total_flow: number
  top_channel: string | null
  top_channel_flow: number
}

export type QuickSdkRdLineListResponse = {
  items: QuickSdkRdLineSuggestion[]
  total: number
}

export type QuickSdkGameFlowResponse = QuickSdkRdLineSuggestion

export type QuickSdkSummaryResponse = {
  batch_count: number
  row_count: number
  game_count: number
  channel_count: number
  total_flow: number | string
}

export type QuickSdkBatch = {
  id: string
  source_file: string
  settlement_month: string
  row_count: number
  game_count: number
  channel_count: number
  total_flow: number | string
  imported_at?: string | null
}

export type QuickSdkBatchListResponse = {
  items: QuickSdkBatch[]
  total: number
}

export type QuickSdkMonthlySummaryRow = {
  settlement_month: string
  row_count: number
  game_count: number
  channel_count: number
  total_flow: number | string
}

export type QuickSdkRankingRow = {
  name: string
  row_count: number
  total_flow: number | string
  share_rate: number
}

export type QuickSdkAnalyticsResponse = {
  monthly: QuickSdkMonthlySummaryRow[]
  game_rankings: QuickSdkRankingRow[]
  channel_rankings: QuickSdkRankingRow[]
}

type QuickSdkFlowRow = {
  settlement_month?: string | null
  game_name?: string | null
  channel_name?: string | null
  gross_flow?: number | string | null
}

type QuickSdkFlowListResponse = {
  items: QuickSdkFlowRow[]
  total: number
}

function queryString(params: Record<string, unknown>): string {
  const q = new URLSearchParams()
  for (const [key, raw] of Object.entries(params)) {
    if (raw === undefined || raw === null) continue
    const value = String(raw).trim()
    if (value) q.set(key, value)
  }
  const qs = q.toString()
  return qs ? `?${qs}` : ''
}

export function listQuickSdkRdLines(params: {
  settlement_month?: string
  q?: string
  limit?: number
}): Promise<QuickSdkRdLineListResponse> {
  return apiGet<QuickSdkRdLineListResponse>(`${PATH}/rd-lines${queryString(params)}`).catch(
    async (error) => {
      if (!(error instanceof ApiError) || error.status !== 404) throw error
      return listQuickSdkRdLinesFromFlows(params)
    }
  )
}

export function getQuickSdkGameFlow(params: {
  settlement_month?: string
  game_name: string
}): Promise<QuickSdkGameFlowResponse> {
  return apiGet<QuickSdkGameFlowResponse>(`${PATH}/game-flow${queryString(params)}`).then(
    async (item) => {
      const flow = Number(item?.total_flow || 0)
      const rowCount = Number(item?.row_count || 0)
      if (Number.isFinite(flow) && flow > 0 && rowCount > 0) return item
      return getQuickSdkGameFlowFromSuggestions(params)
    }
  )
}

export function getQuickSdkSummary(params: {
  settlement_month?: string
}): Promise<QuickSdkSummaryResponse> {
  return apiGet<QuickSdkSummaryResponse>(`${PATH}/summary${queryString(params)}`)
}

export function listQuickSdkBatches(params: {
  settlement_month?: string
  limit?: number
  offset?: number
}): Promise<QuickSdkBatchListResponse> {
  return apiGet<QuickSdkBatchListResponse>(`${PATH}/batches${queryString(params)}`)
}

export function getQuickSdkAnalytics(params: {
  settlement_month?: string
}): Promise<QuickSdkAnalyticsResponse> {
  return apiGet<QuickSdkAnalyticsResponse>(`${PATH}/analytics${queryString(params)}`)
}

async function getQuickSdkGameFlowFromSuggestions(params: {
  settlement_month?: string
  game_name: string
}): Promise<QuickSdkGameFlowResponse> {
  const target = String(params.game_name || '').trim()
  if (!target) {
    return {
      game_name: '',
      settlement_month: params.settlement_month || null,
      row_count: 0,
      channel_count: 0,
      source_game_count: 0,
      total_flow: 0,
      top_channel: null,
      top_channel_flow: 0
    }
  }
  const response = await listQuickSdkRdLines({
    settlement_month: params.settlement_month,
    q: target,
    limit: 500
  })
  const matches = (response.items || []).filter((item) => {
    const name = String(item.game_name || '').trim()
    return name === target || name.includes(target) || normalizeGameName(name) === target
  })
  if (matches.length === 0) {
    return {
      game_name: target,
      settlement_month: params.settlement_month || null,
      row_count: 0,
      channel_count: 0,
      source_game_count: 0,
      total_flow: 0,
      top_channel: null,
      top_channel_flow: 0
    }
  }
  if (matches.length === 1) return matches[0]

  const top = matches
    .slice()
    .sort((a, b) => Number(b.top_channel_flow || 0) - Number(a.top_channel_flow || 0))[0]
  const totalFlow = matches.reduce((sum, item) => sum + Number(item.total_flow || 0), 0)
  return {
    game_name: target,
    settlement_month: params.settlement_month || matches[0].settlement_month || null,
    row_count: matches.reduce((sum, item) => sum + Number(item.row_count || 0), 0),
    channel_count: matches.reduce((sum, item) => sum + Number(item.channel_count || 0), 0),
    source_game_count: matches.reduce((sum, item) => sum + Number(item.source_game_count || 0), 0),
    total_flow: Number(totalFlow.toFixed(2)),
    top_channel: top?.top_channel || null,
    top_channel_flow: Number(top?.top_channel_flow || 0)
  }
}

async function listQuickSdkRdLinesFromFlows(params: {
  settlement_month?: string
  q?: string
  limit?: number
}): Promise<QuickSdkRdLineListResponse> {
  const limit = Math.max(Number(params.limit || 300), 300)
  const response = await apiGet<QuickSdkFlowListResponse>(
    `${PATH}/flows${queryString({ settlement_month: params.settlement_month, limit })}`
  )
  const keyword = String(params.q || '').trim().toLowerCase()
  const groups = new Map<
    string,
    {
      settlement_month: string | null
      row_count: number
      total_flow: number
      source_games: Set<string>
      channels: Map<string, number>
    }
  >()

  for (const row of response.items || []) {
    const sourceGame = String(row.game_name || '').trim()
    if (!sourceGame) continue
    const gameName = normalizeGameName(sourceGame)
    if (keyword && !gameName.toLowerCase().includes(keyword) && !sourceGame.toLowerCase().includes(keyword)) {
      continue
    }
    const channelName = String(row.channel_name || '').trim() || '未填渠道'
    const flow = Number(row.gross_flow || 0)
    const group =
      groups.get(gameName) ||
      {
        settlement_month: row.settlement_month || params.settlement_month || null,
        row_count: 0,
        total_flow: 0,
        source_games: new Set<string>(),
        channels: new Map<string, number>()
      }

    group.row_count += 1
    group.total_flow += Number.isFinite(flow) ? flow : 0
    group.source_games.add(sourceGame)
    group.channels.set(channelName, (group.channels.get(channelName) || 0) + (Number.isFinite(flow) ? flow : 0))
    groups.set(gameName, group)
  }

  const items = Array.from(groups.entries())
    .map(([gameName, group]) => {
      const [topChannel, topChannelFlow] =
        Array.from(group.channels.entries()).sort((a, b) => b[1] - a[1])[0] || [null, 0]
      return {
        game_name: gameName,
        settlement_month: group.settlement_month,
        row_count: group.row_count,
        channel_count: group.channels.size,
        source_game_count: group.source_games.size,
        total_flow: Number(group.total_flow.toFixed(2)),
        top_channel: topChannel,
        top_channel_flow: Number(topChannelFlow.toFixed(2))
      }
    })
    .sort((a, b) => b.total_flow - a.total_flow)

  return {
    items: items.slice(0, params.limit || items.length),
    total: items.length
  }
}

function normalizeGameName(value: string): string {
  return value
    .replace(/005专服\d+.*/u, '')
    .replace(/005折混服.*/u, '')
    .replace(/005$/u, '')
    .trim()
}

File: backend/app/api/quicksdk.py
"""QuickSDK流水库读取 API。"""

from __future__ import annotations

import re
from dataclasses import dataclass, field

from fastapi import APIRouter, Depends, Query
from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.core.deps import get_db
from app.models.quicksdk import QuickSdkFlow, QuickSdkImportBatch
from app.schemas.quicksdk import (
    QuickSdkBatchListResponse,
    QuickSdkBatchRead,
    QuickSdkGameFlowResponse,
    QuickSdkRdLineListResponse,
    QuickSdkRdLineSuggestion,
    QuickSdkSummaryResponse,
)

router = APIRouter()


def _normalize_month(raw: str | None) -> str | None:
    """将 2026年5月 / 2026-5 / 202605 统一为 2026-05。"""
    if raw is None:
        return None
    text = str(raw).strip()
    if not text:
        return None
    match = re.match(r"^(\d{4})年(\d{1,2})月$", text)
    if match:
        return f"{match.group(1)}-{int(match.group(2)):02d}"
    match = re.match(r"^(\d{4})[-/.](\d{1,2})$", text)
    if match:
        return f"{match.group(1)}-{int(match.group(2)):02d}"
    match = re.match(r"^(\d{4})(\d{2})$", text)
    if match:
        return f"{match.group(1)}-{int(match.group(2)):02d}"
    return text


def _normalize_product_name(raw: str | None) -> str:
    """
    QuickSDK 原始游戏名常带包服/折扣后缀。研发账单通常按项目组出单，
    这里只做确定性高的归并，剩余名称保持原样。
    """
    text = str(raw or "").strip()
    if not text:
        return ""
    normalized = re.sub(r"005专服\d+.*$", "", text)
    normalized = re.sub(r"005折混服.*$", "", normalized)
    normalized = re.sub(r"005$", "", normalized)
    normalized = normalized.strip()
    return normalized or text


def _matches_query(summary_name: str, source_names: set[str], query: str | None) -> bool:
    if not query:
        return True
    q = query.strip().lower()
    if not q:
        return True
    if q in summary_name.lower():
        return True
    return any(q in name.lower() for name in source_names)


@dataclass
class _FlowSummary:
    game_name: str
    settlement_month: str | None
    row_count: int = 0
    total_flow: float = 0
    channels: set[str] = field(default_factory=set)
    source_games: set[str] = field(default_factory=set)
    channel_totals: dict[str, float] = field(default_factory=dict)

    def add(self, source_game: str, channel: str, flow: float) -> None:
        self.row_count += 1
        self.total_flow += flow
        if source_game:
            self.source_games.add(source_game)
        if channel:
            self.channels.add(channel)
            self.channel_totals[channel] = self.channel_totals.get(channel, 0) + flow

    def to_response(self) -> QuickSdkRdLineSuggestion:
        top_channel = None
        top_channel_flow = 0.0
        if self.channel_totals:
            top_channel, top_channel_flow = max(
                self.channel_totals.items(), key=lambda item: item[1]
            )
        return QuickSdkRdLineSuggestion(
            game_name=self.game_name,
            settlement_month=self.settlement_month,
            row_count=self.row_count,
            channel_count=len(self.channels),
            source_game_count=len(self.source_games),
            total_flow=round(self.total_flow, 2),
            top_channel=top_channel,
            top_channel_flow=round(top_channel_flow, 2),
        )


def _build_flow_summaries(
    db: Session,
    *,
    settlement_month: str | None,
    q: str | None = None,
) -> list[_FlowSummary]:
    stmt = select(
        QuickSdkFlow.settlement_month,
        QuickSdkFlow.game_name,
        QuickSdkFlow.channel_name,
        QuickSdkFlow.gross_flow,
    )
    if settlement_month:
        stmt = stmt.where(QuickSdkFlow.settlement_month == settlement_month)

    rows = db.execute(stmt).all()
    grouped: dict[tuple[str | None, str], _FlowSummary] = {}
    for month, source_game, channel, gross_flow in rows:
        source_game_text = str(source_game or "").strip()
        product_name = _normalize_product_name(source_game_text)
        if not product_name:
            continue
        key = (month, product_name)
        item = grouped.get(key)
        if item is None:
            item = _FlowSummary(game_name=product_name, settlement_month=month)
            grouped[key] = item
        item.add(source_game_text, str(channel or "").strip(), float(gross_flow or 0))

    summaries = [
        item
        for item in grouped.values()
        if _matches_query(item.game_name, item.source_games, q)
    ]
    summaries.sort(key=lambda item: (-item.total_flow, item.game_name))
    return summaries


@router.get("/batches", response_model=QuickSdkBatchListResponse)
def list_quicksdk_batches(
    db: Session = Depends(get_db),
    settlement_month: str | None = Query(None),
    limit: int = Query(100, ge=1, le=500),
    offset: int = Query(0, ge=0),
) -> QuickSdkBatchListResponse:
    month = _normalize_month(settlement_month)
    stmt = select(QuickSdkImportBatch)
    count_stmt = select(func.count(QuickSdkImportBatch.id))
    if month:
        stmt = stmt.where(QuickSdkImportBatch.settlement_month == month)
        count_stmt = count_stmt.where(QuickSdkImportBatch.settlement_month == month)
    total = int(db.execute(count_stmt).scalar_one())
    rows = (
        db.execute(
            stmt.order_by(QuickSdkImportBatch.imported_at.desc()).limit(limit).offset(offset)
        )
        .scalars()
        .all()
    )
    return QuickSdkBatchListResponse(
        items=[QuickSdkBatchRead.model_validate(row) for row in rows],
        total=total,
    )


@router.get("/summary", response_model=QuickSdkSummaryResponse)
def get_quicksdk_summary(
    db: Session = Depends(get_db),
    settlement_month: str | None = Query(None),
) -> QuickSdkSummaryResponse:
    month = _normalize_month(settlement_month)
    stmt = select(
        func.count(func.distinct(QuickSdkImportBatch.id)),
        func.coalesce(func.sum(QuickSdkImportBatch.row_count), 0),
        func.coalesce(func.sum(QuickSdkImportBatch.total_flow), 0),
    )
    if month:
        stmt = stmt.where(QuickSdkImportBatch.settlement_month == month)
    batch_count, row_count, total_flow = db.execute(stmt).one()

    flow_stmt = select(
        func.count(func.distinct(QuickSdkFlow.game_name)),
        func.count(func.distinct(QuickSdkFlow.channel_name)),
    )
    if month:
        flow_stmt = flow_stmt.where(QuickSdkFlow.settlement_month == month)
    game_count, channel_count = db.execute(flow_stmt).one()
    return QuickSdkSummaryResponse(
        batch_count=int(batch_count or 0),
        row_count=int(row_count or 0),
        game_count=int(game_count or 0),
        channel_count=int(channel_count or 0),
        total_flow=float(total_flow or 0),
    )


@router.get("/rd-lines", response_model=QuickSdkRdLineListResponse)
def list_rd_line_suggestions(
    db: Session = Depends(get_db),
    settlement_month: str | None = Query(None),
    q: str | None = Query(None),
    limit: int = Query(300, ge=1, le=500),
) -> QuickSdkRdLineListResponse:
    month = _normalize_month(settlement_month)
    summaries = _build_flow_summaries(db, settlement_month=month, q=q)
    return QuickSdkRdLineListResponse(
        items=[item.to_response() for item in summaries[:limit]],
        total=len(summaries),
    )


@router.get("/game-flow", response_model=QuickSdkGameFlowResponse)
def get_game_flow(
    db: Session = Depends(get_db),
    game_name: str = Query(...),
    settlement_month: str | None = Query(None),
) -> QuickSdkGameFlowResponse:
    month = _normalize_month(settlement_month)
    target = str(game_name or "").strip()
    summaries = _build_flow_summaries(db, settlement_month=month, q=target)
    selected = next((item for item in summaries if item.game_name == target), None)
    if selected is None:
        selected = next((item for item in summaries if target in item.source_games), None)
    if selected is None:
        selected = summaries[0] if len(summaries) == 1 else None
    if selected is None:
        return QuickSdkGameFlowResponse(
            game_name=target,
            settlement_month=month,
        )
    item = selected.to_response()
    return QuickSdkGameFlowResponse(
        game_name=item.game_name,
        settlement_month=item.settlement_month,
        row_count=item.row_count,
        channel_count=item.channel_count,
        source_game_count=item.source_game_count,
        total_flow=item.total_flow,
        top_channel=item.top_channel,
        top_channel_flow=item.top_channel_flow,
    )
